graphql_compiler/compiler/emit_sql.py

```diff
@@ -0,0 +1,747 @@
+            column = _get_column(selectable, field_name).label(None)
+            columns.append(column)
+            field_renames = context.query_path_field_renames[tag_field.location.query_path]
+            field_renames[field_name] = column.name
+    return columns
+
+
+def _create_query(node, is_final_query, context):
+    """Create a query from a SqlNode.
+
+    If this query is the final query, filters do not need to be applied, and intermediate link
+    columns and tag columns do not need to be included in output.
+    """
+    # filters are computed before output columns, so that tag columns can be resolved before any
+    # renames occur for columns involved in output
+    filter_clauses = []
+    if not is_final_query:
+        filter_clauses = [
+            _convert_filter_to_sql(filter_block, filter_query_path, context)
+            for filter_block, filter_query_path in context.query_path_to_filter[node.query_path]
+        ]
+
+    columns = _get_output_columns(node, is_final_query, context)
+    if not is_final_query:
+        # for every recursion that is a child of this node, include the link column to the child
+        # recursion in this node's query's outputs
+        for recursion in node.recursions:
+            in_col, _ = context.query_path_to_recursion_columns[recursion.query_path]
+            columns.append(in_col)
+    # If this node is completing a recursion, include the outward column in this node's outputs
+    if node.query_path in context.query_path_to_recursion_columns:
+        _, out_col = context.query_path_to_recursion_columns[node.query_path]
+        columns.append(out_col)
+
+    from_clause = _get_node_from_clause(node, context)
+    query = select(columns).select_from(from_clause)
+    if is_final_query:
+        return query
+    return query.where(and_(*filter_clauses))
+
+
+def _convert_filter_to_sql(filter_block, filter_query_path, context):
+    """Return the SQLAlchemy expression for a Filter predicate."""
+    filter_location_info = context.query_path_to_location_info[filter_query_path]
+    filter_selectable = context.query_path_to_selectable[filter_query_path]
+    expression = filter_block.predicate
+    return _expression_to_sql(expression, filter_selectable, filter_location_info, context)
+
+
+def _expression_to_sql(expression, selectable, location_info, context):
+    """Recursively convert a compiler predicate to it's SQLAlchemy expression representation."""
+    if isinstance(expression, expressions.LocalField):
+        column_name = expression.field_name
+        column = _get_column(selectable, column_name)
+        return column
+    if isinstance(expression, expressions.Variable):
+        variable_name = expression.variable_name
+        return bindparam(variable_name)
+    if isinstance(expression, expressions.Literal):
+        return expression.value
+    if isinstance(expression, expressions.ContextField):
+        tag_field_name = expression.location.field
```

Comment:
consider making anything that isn't 1-3 lines' worth of code in the top-level `if` into its own helper function -- that will massively shrink the size of this function and improve readability